File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
	public int[] solution(String[] genres, int[] plays) {
		Map<String, List<Song>> songsByGenre = new HashMap<>();
		Map<String, Integer> totalPlaysByGenre = new HashMap<>();

		for (int i = 0; i < genres.length; i++) {
			songsByGenre.computeIfAbsent(genres[i], (key) -> new ArrayList<>()).add(new Song(i, plays[i]));
			totalPlaysByGenre.merge(genres[i], plays[i], Integer::sum);
		}

		List<Integer> answer = new ArrayList<>();

		totalPlaysByGenre.entrySet()
			.stream()
			.sorted(Comparator.comparingInt((Map.Entry<String, Integer> entry) -> entry.getValue()).reversed())
			.forEach((entry) -> {
				songsByGenre.get(entry.getKey()).stream().sorted(Comparator.comparingInt((Song song) -> song.plays)
						.reversed()
						.thenComparingInt((Song song) -> song.idx))
					.limit(2)
					.forEach(song -> {
						answer.add(song.idx);
					});
			});

		return answer.stream().mapToInt(Integer::intValue).toArray();
	}

	class Song {
		public int idx;
		public int plays;

		public Song(int idx, int plays) {
			this.idx = idx;
			this.plays = plays;
		}
	}
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from typing import List

class Song:
    def __init__(self, id: int, genre: str, play: int):
        self.id = id
        self.play = play
        self.genre = genre

def solution(genres: List[str], plays: List[int]):
    plays_by_genre: dict[str, int] = {}
    
    songs = []
    
    for idx, info in enumerate(zip(genres, plays)):
        genre = info[0]
        play = info[1]
        
        songs.append(Song(idx, genre, play))
        
        if genre in plays_by_genre:
            plays_by_genre[genre] += play
        else:
            plays_by_genre[genre] = play
            
    filtered_songs = sum(
        [sorted(
            [song for song in songs if song.genre == genre], 
            key=lambda song: song.play, 
            reverse=True)[:2] 
         for genre in plays_by_genre.keys()], 
        [])
            
    genre_rank_by_plays = {
        info[0]: rank 
        for rank, info in enumerate(sorted(
            plays_by_genre.items(), 
            key = lambda entry: entry[1], 
            reverse=True))}
    
    return [song.id for song in sorted(
        filtered_songs, 
        key=lambda song: (genre_rank_by_plays[song.genre], -song.play, song.id))]